File: sdks/local-rp/typescript/src/claims.ts
// Claim signing/verification, mirroring `crates/liblinkkeys/src/claims.rs`'s
// subset this SDK needs: verifying claims returned by ticket redemption
// against per-signer-domain public keys. `sign_claim` is included too (small,
// and useful for tests/fakes that need to mint a realistic `Claim`).

import { encodeCborTuple, text, bytes as tupleBytes, textOrNull } from "./cborTuple.ts";
import { signEd25519, verifyEd25519 } from "./crypto.ts";
import type { Claim, ClaimSignature, DomainPublicKey } from "./generated/types.gen.ts";

export type ClaimErrorCode =
  | "signature-invalid"
  | "key-not-found"
  | "key-revoked"
  | "key-expired"
  | "revoked"
  | "expired"
  | "bad-expiry"
  | "unsigned"
  | "domain-keys-unavailable"
  | "domain-unverified";

export class ClaimError extends Error {
  readonly code: ClaimErrorCode;
  readonly detail?: string;
  constructor(code: ClaimErrorCode, detail?: string) {
    super(detail ? `${code}: ${detail}` : code);
    this.name = "ClaimError";
    this.code = code;
    this.detail = detail;
  }
}

const CLAIM_PAYLOAD_TAG = "linkkeys-claim-v2";

/**
 * Canonical bytes a single signature covers for a claim. The subject is the
 * full identity `user_id@subject_domain` (never a bare user_id), and
 * `signing_domain` is bound per-signature, so a signature from one domain
 * can't be relabeled as another's, and a claim about one subject domain
 * can't be replayed against a different one. `claimValue` is a raw byte
 * string (bstr) in the payload, never text — see `claims.json`'s
 * `claim_non_utf8_binary_value` conformance case. Exported so conformance
 * tests can byte-compare against `claims.json`'s `signed_payload_cbor_hex`.
 */
export function claimSignPayload(
  claimId: string,
  claimType: string,
  claimValue: Uint8Array,
  userId: string,
  subjectDomain: string,
  signingDomain: string,
  expiresAt: string | undefined,
  attestedAt: string,
): Uint8Array {
  const subject = `${userId}@${subjectDomain}`;
  return encodeCborTuple([
    text(CLAIM_PAYLOAD_TAG),
    text(claimId),
    text(claimType),
    tupleBytes(claimValue),
    text(subject),
    text(signingDomain),
    textOrNull(expiresAt),
    text(attestedAt),
  ]);
}

export interface ClaimSpec {
  claimId: string;
  claimType: string;
  claimValue: Uint8Array;
  userId: string;
  subjectDomain: string;
  expiresAt?: string;
  attestedAt: string;
}

export interface ClaimSigner {
  domain: string;
  keyId: string;
  privateKeySeed: Uint8Array;
}

/** Sign a claim with one or more keys, producing a `Claim` carrying one `ClaimSignature` per signer. */
export function signClaim(spec: ClaimSpec, signers: readonly ClaimSigner[]): Claim {
  const signatures: ClaimSignature[] = signers.map((signer) => {
    const payload = claimSignPayload(
      spec.claimId,
      spec.claimType,
      spec.claimValue,
      spec.userId,
      spec.subjectDomain,
      signer.domain,
      spec.expiresAt,
      spec.attestedAt,
    );
    return {
      domain: signer.domain,
      signedByKeyId: signer.keyId,
      signature: signEd25519(payload, signer.privateKeySeed),
    };
  });

  return {
    claimId: spec.claimId,
    userId: spec.userId,
    claimType: spec.claimType,
    claimValue: spec.claimValue,
    signatures,
    attestedAt: spec.attestedAt,
    createdAt: new Date().toISOString(),
    expiresAt: spec.expiresAt,
    revokedAt: undefined,
  };
}

/** A domain and the set of its currently-known public keys, as supplied to `verifyClaim`. */
export interface DomainKeySet {
  domain: string;
  keys: readonly DomainPublicKey[];
}

function signingKeyValidity(key: DomainPublicKey, now: Date): "valid" | "revoked" | "expired" {
  if (key.revokedAt !== undefined) return "revoked";
  const expires = new Date(key.expiresAt);
  if (Number.isNaN(expires.getTime())) return "expired";
  return now.getTime() > expires.getTime() ? "expired" : "valid";
}

function verifyOneSignature(
  sig: ClaimSignature,
  payload: Uint8Array,
  keys: readonly DomainPublicKey[],
  now: Date,
): void {
  const key = keys.find((k) => k.keyId === sig.signedByKeyId);
  if (!key) {
    throw new ClaimError("key-not-found", sig.signedByKeyId);
  }
  // A claim signature must come from a signing key, never an encryption key
  // sharing the same id.
  if (key.keyUsage !== "sign") {
    throw new ClaimError("signature-invalid");
  }
  const validity = signingKeyValidity(key, now);
  if (validity === "revoked") throw new ClaimError("key-revoked", key.keyId);
  if (validity === "expired") throw new ClaimError("key-expired", key.keyId);

  if (key.algorithm !== "ed25519") {
    throw new ClaimError("signature-invalid", `unsupported algorithm: ${key.algorithm}`);
  }
  if (!verifyEd25519(payload, sig.signature, key.publicKey)) {
    throw new ClaimError("signature-invalid");
  }
}

/**
 * The cryptographic per-domain quorum: EVERY distinct domain that signed
 * must contribute at least one signature from a currently-valid key of that
 * domain, over `payloadFor(signingDomain)`.
 */
function verifySignatureQuorum(
  signatures: readonly ClaimSignature[],
  domainKeys: readonly DomainKeySet[],
  payloadFor: (signingDomain: string) => Uint8Array,
  now: Date,
): void {
  if (signatures.length === 0) {
    throw new ClaimError("unsigned");
  }

  const domains = [...new Set(signatures.map((s) => s.domain))].sort();

  for (const signingDomain of domains) {
    const set = domainKeys.find((s) => s.domain === signingDomain);
    if (!set) {
      throw new ClaimError("domain-keys-unavailable", signingDomain);
    }
    const payload = payloadFor(signingDomain);

    let lastError: ClaimError = new ClaimError("domain-unverified", signingDomain);
    let satisfied = false;
    for (const sig of signatures.filter((s) => s.domain === signingDomain)) {
      try {
        verifyOneSignature(sig, payload, set.keys, now);
        satisfied = true;
        break;
      } catch (e) {
        if (e instanceof ClaimError) lastError = e;
        else throw e;
      }
    }
    if (!satisfied) throw lastError;
  }
}

/** Verify only the cryptographic per-domain quorum (not revocation/expiry — see `verifyClaim`). */
export function verifyClaimSignatures(
  claim: Claim,
  subjectDomain: string,
  domainKeys: readonly DomainKeySet[],
  now: Date = new Date(),
): void {
  verifySignatureQuorum(
    claim.signatures,
    domainKeys,
    (signingDomain) =>
      claimSignPayload(
        claim.claimId,
        claim.claimType,
        claim.claimValue,
        claim.userId,
        subjectDomain,
        signingDomain,
        claim.expiresAt,
        claim.attestedAt,
      ),
    now,
  );
}

/** Full claim verification: the cryptographic quorum plus the claim's own revocation and expiry. */
export function verifyClaim(
  claim: Claim,
  subjectDomain: string,
  domainKeys: readonly DomainKeySet[],
  now: Date = new Date(),
): void {
  verifyClaimSignatures(claim, subjectDomain, domainKeys, now);

  if (claim.revokedAt !== undefined) {
    throw new ClaimError("revoked");
  }
  if (claim.expiresAt !== undefined) {
    const expires = new Date(claim.expiresAt);
    if (Number.isNaN(expires.getTime())) {
      throw new ClaimError("bad-expiry");
    }
    if (now.getTime() > expires.getTime()) {
      throw new ClaimError("expired");
    }
  }
}

File: sdks/local-rp/elixir/lib/linkkeys_local_rp/claims.ex
defmodule LinkkeysLocalRp.Claims do
  @moduledoc """
  Claim signature/revocation/expiry verification.

  Mirrors `crates/liblinkkeys/src/claims.rs` for exactly the pieces
  `complete_local_login` needs: per-signer-domain signature quorum,
  revocation, and expiry. `sign_claim/2` is included only so this
  package's own flow tests can build fake claims exactly like
  `sdks/local-rp/rust/tests/flow.rs` does (IDP-side operation; the SDK
  itself only ever verifies claims returned from a ticket redemption,
  never signs them).
  """

  alias LinkkeysLocalRp.Cbor
  alias LinkkeysLocalRp.Crypto
  alias LinkkeysLocalRp.Timeutil
  alias LinkkeysLocalRp.Types.{Claim, ClaimSignature}

  @claim_payload_tag "linkkeys-claim-v2"

  defmodule DomainKeySet do
    @moduledoc "A signer domain and the (already DNS-pinned) public keys fetched for it."
    defstruct [:domain, :keys]
  end

  defmodule ClaimSpec do
    @moduledoc "Test-support input to `sign_claim/2` (IDP-side operation)."
    defstruct [
      :claim_id,
      :claim_type,
      :claim_value,
      :user_id,
      :subject_domain,
      :attested_at,
      expires_at: nil
    ]
  end

  defmodule ClaimSigner do
    @moduledoc "Test-support signer input to `sign_claim/2`."
    defstruct [:domain, :key_id, :private_key]
  end

  defp claim_sign_payload(
         claim_id,
         claim_type,
         claim_value,
         user_id,
         subject_domain,
         signing_domain,
         expires_at,
         attested_at
       ) do
    # The subject is bound as the single full identity `user_id@subject_domain`
    # (not the bare user_id), so a claim about a user_id at one domain can't
    # be replayed as the same user_id at another. `signing_domain` — the
    # attestor for THIS signature — is bound per-signature.
    subject = "#{user_id}@#{subject_domain}"

    Cbor.encode([
      @claim_payload_tag,
      claim_id,
      claim_type,
      Cbor.bytes(claim_value),
      subject,
      signing_domain,
      expires_at,
      attested_at
    ])
  end

  @doc "Sign a claim with one or more keys, producing a `Claim` carrying one `ClaimSignature` per signer. IDP-side operation; see module docs."
  def sign_claim(%ClaimSpec{} = spec, signers) when is_list(signers) do
    signatures =
      Enum.map(signers, fn %ClaimSigner{} = signer ->
        payload =
          claim_sign_payload(
            spec.claim_id,
            spec.claim_type,
            spec.claim_value,
            spec.user_id,
            spec.subject_domain,
            signer.domain,
            spec.expires_at,
            spec.attested_at
          )

        signature = Crypto.ed25519_sign(payload, signer.private_key)
        %ClaimSignature{domain: signer.domain, signed_by_key_id: signer.key_id, signature: signature}
      end)

    %Claim{
      claim_id: spec.claim_id,
      user_id: spec.user_id,
      claim_type: spec.claim_type,
      claim_value: spec.claim_value,
      signatures: signatures,
      attested_at: spec.attested_at,
      created_at: spec.attested_at,
      expires_at: spec.expires_at,
      revoked_at: nil
    }
  end

  defp verify_one_signature(%ClaimSignature{} = sig, payload, keys, now) do
    case Enum.find(keys, fn k -> k.key_id == sig.signed_by_key_id end) do
      nil ->
        {:error, {:key_not_found, sig.signed_by_key_id}}

      key ->
        cond do
          key.key_usage != "sign" ->
            {:error, :signature_invalid}

          # Gates the SIGNING KEY's own revocation/expiry (not the claim's,
          # which verify_claim/4 checks separately). `now` is threaded
          # through explicitly (this SDK never reads the system clock
          # internally), which is strictly more testable/deterministic
          # than liblinkkeys' own key-validity check (which reads
          # Utc::now() directly at this exact point, a documented
          # exception to its "explicit now" discipline).
          key.revoked_at != nil ->
            {:error, {:key_revoked, key.key_id}}

          true ->
            case key_valid_at(key, now) do
              :ok ->
                if Crypto.ed25519_verify(payload, sig.signature, key.public_key),
                  do: :ok,
                  else: {:error, :signature_invalid}

              error ->
                error
            end
        end
    end
  end

  defp key_valid_at(key, now) do
    case safe_parse(key.expires_at) do
      {:ok, expires} ->
        if DateTime.compare(now, expires) == :gt,
          do: {:error, {:key_expired, key.key_id}},
          else: :ok

      {:error, _} ->
        {:error, {:key_expired, key.key_id}}
    end
  end

  defp safe_parse(s) do
    {:ok, Timeutil.parse_rfc3339(s)}
  rescue
    _ -> {:error, :bad_timestamp}
  end

  @doc "Every distinct domain that signed must contribute at least one signature from a currently-valid key of that domain."
  def verify_claim_signatures(%Claim{} = claim, subject_domain, domain_keys, now) do
    if claim.signatures == [] do
      {:error, :unsigned}
    else
      domains = claim.signatures |> Enum.map(& &1.domain) |> Enum.uniq() |> Enum.sort()
      do_verify_domains(domains, claim, subject_domain, domain_keys, now)
    end
  end

  defp do_verify_domains([], _claim, _subject_domain, _domain_keys, _now), do: :ok

  defp do_verify_domains([signing_domain | rest], claim, subject_domain, domain_keys, now) do
    case Enum.find(domain_keys, fn s -> s.domain == signing_domain end) do
      nil ->
        {:error, {:domain_keys_unavailable, signing_domain}}

      %DomainKeySet{keys: keys} ->
        payload =
          claim_sign_payload(
            claim.claim_id,
            claim.claim_type,
            claim.claim_value,
            claim.user_id,
            subject_domain,
            signing_domain,
            claim.expires_at,
            claim.attested_at
          )

        sigs_for_domain = Enum.filter(claim.signatures, fn s -> s.domain == signing_domain end)

        case verify_any(sigs_for_domain, payload, keys, now) do
          :ok -> do_verify_domains(rest, claim, subject_domain, domain_keys, now)
          error -> error
        end
    end
  end

  defp verify_any([], _payload, _keys, _now), do: {:error, {:domain_unverified, nil}}

  defp verify_any([sig | rest], payload, keys, now) do
    case verify_one_signature(sig, payload, keys, now) do
      :ok -> :ok
      {:error, _} = err -> if rest == [], do: err, else: verify_any(rest, payload, keys, now)
    end
  end

  @doc "Full claim verification: the cryptographic per-domain quorum plus the claim's own revocation and expiry. All must pass."
  def verify_claim(%Claim{} = claim, subject_domain, domain_keys, now) do
    with :ok <- verify_claim_signatures(claim, subject_domain, domain_keys, now) do
      cond do
        claim.revoked_at != nil ->
          {:error, :revoked}

        claim.expires_at != nil ->
          case safe_parse(claim.expires_at) do
            {:ok, expires} ->
              if DateTime.compare(now, expires) == :gt, do: {:error, :expired}, else: :ok

            {:error, _} ->
              {:error, :bad_expiry}
          end

        true ->
          :ok
      end
    end
  end
end
